07-reproj.qmd

```diff
@@ -175,9 +184,106 @@ axes[0].set_title('Original (EPSG:26912)')
+
+First, we create a connection to our "template" raster to read its metadata:
+
+```{python}
+template = rasterio.open('output/nlcd_4326.tif')
+template.meta
+```
+
+Then, we create a write-mode connection to our destination raster, using this metadata, meaning that as the resampling result is going to have identical metadata as the "template":
+
+```{python}
+dst_nlcd_2 = rasterio.open('output/nlcd_4326_2.tif', 'w', **template.meta)
+```
+
+Now, we can resample and write the result:
+
+```{python}
+rasterio.warp.reproject(
+    source=rasterio.band(src_nlcd, 1),
+    destination=rasterio.band(dst_nlcd_2, 1),
+    src_transform=src_nlcd.transform,
+    src_crs=src_nlcd.crs,
+    dst_transform=dst_nlcd_2.transform,
+    dst_crs=dst_nlcd_2.crs,
+    resampling=rasterio.warp.Resampling.nearest
+)
+```
+
+```{python}
+dst_nlcd_2.close()
+```
+
+Naturally, in this case, the outputs `nlcd_4326.tif` and `nlcd_4326_2.tif` are identical, as we used the same "template" and the same source data: 
+
+```{python}
+d = rasterio.open('output/nlcd_4326.tif').read(1) == rasterio.open('output/nlcd_4326_2.tif').read(1)
+d
+```
+
+```{python}
+np.all(d)
+```
 
+The difference is that in the first example we calculate the template automatically, using `rasterio.warp.calculate_default_transform`, while in the second example we used an existing raster as the "template".
 
+Importantly, when the template raster has much more "coarse" resolution than the source raster, the:
+
+* `rasterio.warp.Resampling.average` (for continuous rasters), or 
+* `rasterio.warp.Resampling.mode` (for categorical rasters)
+
+resampling method should be used, instead of `rasterio.warp.Resampling.nearest`. Otherwise, much of the data will be lost, as the "nearest" method can capture one pixel value only for each destination raster pixel. 
+
+Reprojecting continuous rasters (with numeric or, in this case, integer values) follows an almost identical procedure. This is demonstrated below with `srtm.tif` from the Shuttle Radar Topography Mission (SRTM), which represents height in meters above sea level (elevation) with the WGS84 CRS.
+
+We will reproject this dataset into a projected CRS, but not with the nearest neighbor method which is appropriate for categorical data. Instead, we will use the bilinear method which computes the output cell value based on the four nearest cells in the original raster. The values in the projected dataset are the distance-weighted average of the values from these four cells: the closer the input cell is to the center of the output cell, the greater its weight. The following code section create a text string representing WGS 84 / UTM zone 12N, and reproject the raster into this CRS, using the bilinear method. The code is practically the same, except for changing the source and destination file names, and replacing `nearest` with `bilinear`:
```

Comment:
Bilinear is usually preferable no? If so bilinear example could go first is my thought.

Reply:
Not sure I understand what you meant by "first". There is just one example of each: "nearest" for NLCD (categorical) and "bilinear" for SRTM (continuous)